File: python/src/video_player.py
"""A video player class."""

from .video_library import VideoLibrary
from .video_library import VideoLibrary
from .video_playlist import Playlist
from random import choice

class VideoPlayer:
    """A class used to represent a Video Player."""

    def __init__(self):
        self._video_library = VideoLibrary()
        self._playlist = Playlist()
        self._stack = []
        self._play = False
        self._pause = False
        self._flag = {}

    def number_of_videos(self):
        num_videos = len(self._video_library.get_all_videos())
        print(f"{num_videos} videos in the library")

    def number_of_videos(self):
        num_videos = len(self._video_library.get_all_videos())
        print(f"{num_videos} videos in the library")

    def _standard_form(self, video):
        """Returns video description & details in specified form"""
        title = video.title
        url = video.video_id 
        tags = video.tags

        tag_string = ""
        for i, tag in enumerate(tags):
            if i == 0:
                tag_string += tag
            else:
                tag_string += " " + tag

        return (str(title) + ' (' + str(url) + ') ' + '[' + tag_string  + ']')         

    def _get_tagstring(self, video):
        """Returns list of tags in lowercase"""
        tags = video.tags

        tag_string = ""
        for i, tag in enumerate(tags):
            if i == 0:
                tag_string += tag.lower()
            else:
                tag_string += " " + tag.lower()
        return tag_string.split()  

    def _non_flag(self):
        """Returns a list of non-flagged videos"""
        non = []
        for vid in self._video_library.get_all_videos():   
            if vid not in self._flag:
                non.append(vid)
        return non 

    def show_all_videos(self):
        """Returns a listing of all videos."""

        print("Here's a list of all available videos:")
        for video in self._video_library.get_sorted_videos():
            if video in self._flag:
                print(f"{self._standard_form(video)} - FLAGGED {self._flag[video]}")
            else:    
                print(self._standard_form(video))
            
    def play_video(self, video_id):
        """Plays the respective video.

        Args:
            video_id: The video_id to be played.
        """
        curr_vid = self._video_library.get_video(video_id)

        if curr_vid in self._flag:
            print(f"Cannot play video: Video is currently flagged {self._flag[curr_vid]}")
            return
        if not self._stack and curr_vid:
            print(f"Playing video: {curr_vid.title}")
            self._stack.append(curr_vid)
            self._play = True
            self._pause = False
        elif self._stack and curr_vid:
            print(f"Stopping video: {self._stack.pop().title}")
            print(f"Playing video: {curr_vid.title}")
            self._stack.append(curr_vid) 
            self._play = True
            self._pause = False
        elif not curr_vid:
            print("Cannot play video: Video does not exist")       

    def stop_video(self):
        """Stops the current video."""

        if not self._stack:
            print("Cannot stop video: No video is currently playing")
        else:
            print(f"Stopping video: {self._stack.pop().title}")
            self._play = False
            self._pause = False

    def play_random_video(self):
        """Plays a random video from the video library."""
        if not self._non_flag():
            print("No videos available") 
            return 

        video = choice(self._non_flag())
        if not self._stack:
            print(f"Playing video: {video.title}")
            self._stack.append(video)
            self._play = True
            self._pause = False
        else:
            print(f"Stopping video: {self._stack.pop().title}")  
            print(f"Playing video: {video.title}") 
            self._stack.append(video)    
            self._play = True
            self._pause = False

    def pause_video(self):
        """Pauses the current video."""
        if self._play and not self._pause:
            print(f"Pausing video: {self._stack[-1].title}")
            self._pause = True
        elif self._play and self._pause:
            print(f"Video already paused: {self._stack[-1].title}")    
        elif not self._play:
            print("Cannot pause video: No video is currently playing")        

    def continue_video(self):
        """Resumes playing the current video."""
        if self._play and not self._pause:
            print("Cannot continue video: Video is not paused")
        elif self._pause:
            print(f"Continuing video: {self._stack[-1].title}") 
        else:
            print("Cannot continue video: No video is currently playing")   

    def show_playing(self):
        """Displays video currently playing."""

        if self._play and not self._pause:
            print(f"Currently playing: {self._standard_form(self._stack[-1])}")
        elif self._pause:
            print(f"Currently playing: {self._standard_form(self._stack[-1])} - PAUSED")    
        else:
            print("No video is currently playing")

    def create_playlist(self, playlist_name):
        """Creates a playlist with a given name.

        Args:
            playlist_name: The playlist name.
        """
        self._playlist.create(playlist_name)

    def add_to_playlist(self, playlist_name, video_id):
        """Adds a video to a playlist with a given name.

        Args:
            playlist_name: The playlist name.
            video_id: The video_id to be added.
        """
        self._playlist.add(playlist_name, self._video_library.get_video(video_id), self._flag)

    def show_all_playlists(self):
        """Display all playlists."""

        self._playlist.show_all()

    def show_playlist(self, playlist_name):
        """Display all videos in a playlist with a given name.

        Args:
            playlist_name: The playlist name.
        """
        for list in self._playlist.lists:
            if list.lower() == playlist_name.lower():
                print(f"Showing playlist: {playlist_name}")
                if not self._playlist.lists[list]:
                    print("No videos here yet")
                    return
                else:
                    for video in self._playlist.lists[list]:
                        if video in self._flag:
                            print(f"{self._standard_form(video)} - FLAGGED {self._flag[video]}")
                        else:    
                            print(self._standard_form(video))
                    return

        print("Cannot show playlist another_playlist: Playlist does not exist")  

    def remove_from_playlist(self, playlist_name, video_id):
        """Removes a video to a playlist with a given name.

        Args:
            playlist_name: The playlist name.
            video_id: The video_id to be removed.
        """
        self._playlist.remove(playlist_name, self._video_library.get_video(video_id))

    def clear_playlist(self, playlist_name):
        """Removes all videos from a playlist with a given name.

        Args:
            playlist_name: The playlist name.
        """
        self._playlist.clear(playlist_name)

    def delete_playlist(self, playlist_name):
        """Deletes a playlist with a given name.

        Args:
            playlist_name: The playlist name.
        """
        self._playlist.delete(playlist_name)

    def search_videos(self, search_term):
        """Display all the videos whose titles contain the search_term.

        Args:
            search_term: The query to be used in search.
        """
        cont = []
        for vid in self._non_flag():
            if search_term.lower() in vid.title.lower():
                cont.append(vid.video_id)

        if not cont:
            print(f"No search results for {search_term}")
        else:
            cont.sort()
            print(f"Here are the results for {search_term}:")
            for i,vid_id in enumerate(cont):
                print(f"{i + 1}) {self._standard_form(self._video_library.get_video(vid_id))}")

            print("Would you like to play any of the above? If yes, specify the number of the video.") 
            print("If your answer is not a valid number, we will assume it's a no.")   
            number = input("")
            try:
                if 1 <= int(number) <= len(cont):
                    print(f"Playing video: {self._video_library.get_video(cont[int(number) - 1]).title}") 
            except:
                return 

    def search_videos_tag(self, video_tag):
        """Display all videos whose tags contains the provided tag.

        Args:
            video_tag: The video tag to be used in search.
        """
        cont = []
        for vid in self._non_flag():
            if video_tag.lower() in self._get_tagstring(vid):
                cont.append(vid.video_id)

        if not cont:
            print(f"No search results for {video_tag}")
        else:
            cont.sort()
            print(f"Here are the results for {video_tag}:")
            for i,vid_id in enumerate(cont):
                print(f"{i + 1}) {self._standard_form(self._video_library.get_video(vid_id))}")

            print("Would you like to play any of the above? If yes, specify the number of the video.") 
            print("If your answer is not a valid number, we will assume it's a no.")   
            number = input("")
            try:
                if 1 <= int(number) <= len(cont):
                    print(f"Playing video: {self._video_library.get_video(cont[int(number) - 1]).title}") 
            except:
                return        

    def flag_video(self, video_id, flag_reason=""):
        """Mark a video as flagged.

        Args:
            video_id: The video_id to be flagged.
            flag_reason: Reason for flagging the video.
        """
        if not self._video_library.get_video(video_id):
            print("Cannot flag video: Video does not exist")
            return 
        elif (self._pause or self._play) and self._stack[-1] == self._video_library.get_video(video_id):
            print(f"Stopping video: {self._video_library.get_video(video_id).title}") 
            print(f"Successfully flagged video: {self._video_library.get_video(video_id).title} (reason: {flag_reason})")
            self._flag[self._video_library.get_video(video_id)] = "(reason: " + flag_reason  + ")" 
            self._stack.pop()
            self._play = False
            self._pause = False
            return
        elif (self._pause or self._play) and self._stack[-1] == self._video_library.get_video(video_id):
            print(f"Stopping video: {self._video_library.get_video(video_id).title}") 
            print(f"Successfully flagged video: {self._video_library.get_video(video_id).title} (reason: Not supplied)")
            self._flag[self._video_library.get_video(video_id)] = "(reason: Not supplied)" 
            self._stack.pop()
            self._play = False
            self._pause = False
            return    
        elif not self._flag or not self._video_library.get_video(video_id) in self._flag:
            if not flag_reason:
                print(f"Successfully flagged video: {self._video_library.get_video(video_id).title} (reason: Not supplied)") 
                self._flag[self._video_library.get_video(video_id)] = "(reason: Not supplied)"
            else:
                print(f"Successfully flagged video: {self._video_library.get_video(video_id).title} (reason: {flag_reason})")
                self._flag[self._video_library.get_video(video_id)] = "(reason: " + flag_reason  + ")"  
        else: 
            print("Cannot flag video: Video is already flagged")

    def allow_video(self, video_id):
        """Removes a flag from a video.

        Args:
            video_id: The video_id to be allowed again.
        """
        if not self._video_library.get_video(video_id):
            print("Cannot remove flag from video: Video does not exist")
            return 
        elif self._video_library.get_video(video_id) in self._flag:
            print(f"Successfully removed flag from video: {self._video_library.get_video(video_id).title}") 
            del self._flag[self._video_library.get_video(video_id)]
        else:
            print("Cannot remove flag from video: Video is not flagged") 

File: solutions/java/solution3/src/main/java/com/google/VideoPlayer.java
package com.google;

import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Random;
import java.util.Scanner;
import java.util.stream.Collectors;

public class VideoPlayer {

  enum VideoPlayerState {
    PLAYING,
    PAUSED,
    IDLE
  }

  private final VideoLibrary videoLibrary;
  private Video currentlyPlayingVideo;
  private VideoPlayerState state;

  private HashMap<String, VideoPlaylist> playlists = new HashMap<>();

  public VideoPlayer() {
    // The VideoPlayer constructor
    currentlyPlayingVideo = null;
    state = VideoPlayerState.IDLE;
    this.videoLibrary = new VideoLibrary();
  }

  public void numberOfVideos() {
    System.out.printf("%s videos in the library%n", videoLibrary.getVideos().size());
  }

  public void showAllVideos() {

    var allVideos = videoLibrary.getVideos();

    System.out.println("Here's a list of all available videos:");

    allVideos.stream()
        .sorted(Comparator.comparing(Video::getTitle))
        .collect(Collectors.toList())
        .forEach(video -> System.out.println(video));
  }

  public void playVideo(String videoId) {
    Video v = videoLibrary.getVideo(videoId);

    if(v==null) {
      System.out.println("Cannot play video: Video does not exist");
      return;
    }

    if(v.isFlagged()) {
      StringBuilder reasonString = new StringBuilder();
      reasonString.append("Cannot play video: Video is currently flagged");
      reasonString.append(" "+v.getFormattedReason());
      System.out.println(reasonString);
      return;
    }

    if (currentlyPlayingVideo != null) {
      System.out.println("Stopping video: " + currentlyPlayingVideo.getTitle());
    }
    System.out.println("Playing video: " + v.getTitle());
    currentlyPlayingVideo = v;
    state = VideoPlayerState.PLAYING;
  }

  public void stopVideo() {
    if(currentlyPlayingVideo!=null) {
      System.out.println("Stopping video: "+ currentlyPlayingVideo.getTitle());
      currentlyPlayingVideo = null;
      state = VideoPlayerState.IDLE;
    } else {
      System.out.println("Cannot stop video: No video is currently playing");
    }

  }

  public void playRandomVideo() {
    if(currentlyPlayingVideo!=null) {
      stopVideo();
    }
    Random r = new Random();
    var videos = videoLibrary.getVideos().stream().filter(x -> !x.isFlagged()).collect(Collectors.toList());
    if(videos.size()==0) {
      System.out.println("No videos available");
    } else {
      playVideo(videos.get(r.nextInt(videos.size())).getVideoId());
    }
  }

  public void pauseVideo() {
    switch (state) {
      case PLAYING:
        System.out.println("Pausing video: "+ currentlyPlayingVideo.getTitle());
        state = VideoPlayerState.PAUSED;
        break;
      case PAUSED:
        System.out.println("Video already paused: "+ currentlyPlayingVideo.getTitle());
        break;
      case IDLE:
        System.out.println("Cannot pause video: No video is currently playing");
        break;
      default:
        System.out.println("Unknown State Found!");
        break;
    }
  }

  public void continueVideo() {
    switch (state) {
      case PLAYING:
        System.out.println("Cannot continue video: Video is not paused");
        break;
      case PAUSED:
        System.out.println("Continuing video: "+ currentlyPlayingVideo.getTitle());
        break;
      case IDLE:
        System.out.println("Cannot continue video: No video is currently playing");
        break;
      default:
        System.out.println("Unknown State Found!");
        break;
    }
  }

  public void showPlaying() {
    if(currentlyPlayingVideo!=null) {
      StringBuilder currentlyPlayingString = new StringBuilder();
      currentlyPlayingString.append("Currently playing: "+currentlyPlayingVideo);
      if(state == VideoPlayerState.PAUSED) {
        currentlyPlayingString.append(" - PAUSED");
      }
      System.out.println(currentlyPlayingString);
    } else {
      System.out.println("No video is currently playing");
    }
  }

  public void createPlaylist(String playlistName) {
    String playlistLookupKey = playlistName.toLowerCase(Locale.ROOT);
    if(playlists.containsKey(playlistLookupKey)) {
      System.out.println("Cannot create playlist: A playlist with the same name already exists");
    } else {
      VideoPlaylist newPlaylist = new VideoPlaylist(playlistName);
      System.out.println("Successfully created new playlist: " +playlistName);
      playlists.put(playlistLookupKey, newPlaylist);
    }
  }

  public void addVideoToPlaylist(String playlistName, String videoId) {
    String playlistLookupKey = playlistName.toLowerCase(Locale.ROOT);
    Video videoToAdd = videoLibrary.getVideo(videoId);

    if(!playlists.containsKey(playlistLookupKey)) {
      System.out.println("Cannot add video to "+playlistName+": Playlist does not exist");
      return;
    }

    if(videoToAdd==null) {
      System.out.println("Cannot add video to "+playlistName+": Video does not exist");
      return;
    }

    if(videoToAdd.isFlagged()) {
      StringBuilder reasonString = new StringBuilder();
      reasonString.append("Cannot add video to ");
      reasonString.append(playlistName);
      reasonString.append(": Video is currently flagged (reason: ");
      if(videoToAdd.getFlaggedReason()==null) {
        reasonString.append("Not supplied");
      } else {
        reasonString.append(videoToAdd.getFlaggedReason());
      }
      reasonString.append(")");
      System.out.println(reasonString);
      return;
    }

    VideoPlaylist playlist = playlists.get(playlistLookupKey);

    if(playlist.containsVideo(videoId)) {
      System.out.println("Cannot add video to "+playlistName+": Video already added");
      return;
    } else {
      playlist.addVideo(videoToAdd);
      System.out.println("Added video to "+playlistName+": " +videoToAdd.getTitle());
    }

  }

  public void showPlaylist(String playlistName) {
    String playlistLookupKey = playlistName.toLowerCase(Locale.ROOT);
    if (!playlists.containsKey(playlistLookupKey)) {
      System.out.println("Cannot show playlist "+playlistName+": Playlist does not exist");
    } else {
      var playlist = playlists.get(playlistLookupKey);
      System.out.println("Showing playlist: "+playlistName);
      if (playlist.getVideos().size() == 0) {
        System.out.println("No videos here yet");
      }
      for(Video v : playlist.getVideos()) {
        System.out.println(v);
      }
    }
  }
  public void showAllPlaylists() {
    if (playlists.size() == 0) {
      System.out.println("No playlists exist yet");
      return;
    } else {
      System.out.println("Showing all playlists:");
    }

    playlists.entrySet().stream()
        .map(Map.Entry::getValue)
        .map(v -> v.getName())
        .forEach(name -> System.out.println(name));
  }

  public void removeFromPlaylist(String playlistName, String videoId) {
    String playlistLookupKey = playlistName.toLowerCase(Locale.ROOT);

    if (!playlists.containsKey(playlistLookupKey)) {
      System.out.println("Cannot remove video from "+playlistName+": Playlist does not exist");
    } else {
      var playlist = playlists.get(playlistLookupKey);

      Video videoToRemove = videoLibrary.getVideo(videoId);
      if(videoToRemove==null) {
        System.out.println("Cannot remove video from "+playlistName+": Video does not exist");
        return;
      }

      if(playlist.removeVideo(videoToRemove)) {
        System.out.println("Removed video from "+playlistName+": " + videoToRemove.getTitle());
      } else {
        System.out.println("Cannot remove video from "+playlistName+": Video is not in playlist");
      }
    }
  }

  public void clearPlaylist(String playlistName) {
    String playlistLookupKey = playlistName.toLowerCase(Locale.ROOT);
    if (!playlists.containsKey(playlistLookupKey)) {
      System.out.println("Cannot clear playlist "+playlistName+": Playlist does not exist");
    } else {
      var playlist = playlists.get(playlistLookupKey);
      playlist.clear();
      System.out.println("Successfully removed all videos from "+playlistName);
    }
  }

  public void deletePlaylist(String playlistName) {
    String playlistLookupKey = playlistName.toLowerCase(Locale.ROOT);
    if (playlists.containsKey(playlistLookupKey)) {
      playlists.remove(playlistLookupKey);
      System.out.println("Deleted playlist: " +playlistName);
    } else {
      System.out.println("Cannot delete playlist "+playlistName+": Playlist does not exist");
    }
  }

  public void displaySearchResults(String searchTerm, List<Video> results) {
    if(results.size()==0) {
      System.out.println("No search results for "+searchTerm);
    } else {
      System.out.println("Here are the results for "+searchTerm+":");
      int count = 1;
      for(Video v : results) {
        System.out.println(count + ") " + v);
        count++;
      }
      System.out.println("Would you like to play any of the above? If yes, specify the number of the video. \n"
          + "If your answer is not a valid number, we will assume it's a no.");
      var scanner = new Scanner(System.in);
      if(scanner.hasNextInt()) {
        var option = scanner.nextInt();
        if(results.size()>=option) {
          Video videoToPlay = results.get(option-1);
          playVideo(videoToPlay.getVideoId());
        }
      }
    }
  }

  public void searchVideos(String searchTerm) {
    List<Video> results = videoLibrary.getVideos().stream()
        .filter(v -> v.getTitle().toLowerCase(Locale.ROOT).contains(searchTerm.toLowerCase(Locale.ROOT)))
        .filter(v -> !v.isFlagged())
        .sorted(Comparator.comparing(Video::getTitle))
        .collect(Collectors.toList());

    displaySearchResults(searchTerm, results);
  }

  public void searchVideosWithTag(String videoTag) {
    List<Video> results = videoLibrary.getVideos().stream()
        .filter(v -> v.getTags().toString().toLowerCase(Locale.ROOT).contains(videoTag.toLowerCase(Locale.ROOT)))
        .filter(v -> !v.isFlagged())
        .sorted(Comparator.comparing(Video::getTitle))
        .collect(Collectors.toList());

    displaySearchResults(videoTag, results);
  }

  public void flagVideo(String videoId) {
    flagVideo(videoId, null);
  }

  public void flagVideo(String videoId, String reason) {
    Video videoToFlag = videoLibrary.getVideo(videoId);

    if(videoToFlag==null) {
      System.out.println("Cannot flag video: Video does not exist");
      return;
    }
    if(videoToFlag.isFlagged()) {
      System.out.println("Cannot flag video: Video is already flagged");
      return;
    }

    if(currentlyPlayingVideo!=null && currentlyPlayingVideo.getVideoId().equals(videoId)) {
      stopVideo();
    }

    videoToFlag.markFlagged(reason);

    StringBuilder reasonString = new StringBuilder();
    reasonString.append("Successfully flagged video: ");
    reasonString.append(videoToFlag.getTitle());
    reasonString.append(" "+videoToFlag.getFormattedReason());
    System.out.println(reasonString);
  }

  public void allowVideo(String videoId) {
    Video videoToUnFlag = videoLibrary.getVideo(videoId);

    if(videoToUnFlag==null) {
      System.out.println("Cannot remove flag from video: Video does not exist");
      return;
    }
    if(!videoToUnFlag.isFlagged()) {
      System.out.println("Cannot remove flag from video: Video is not flagged");
      return;
    }
    videoToUnFlag.unFlag();
    System.out.println("Successfully removed flag from video: " + videoToUnFlag.getTitle());
  }
}